cli/determined_cli/tunnel.py

```diff
@@ -0,0 +1,100 @@
+import http.client
+import os
+import socket
+import sys
+import threading
+import urllib
+from typing import Tuple
+
+
+"""
+tunnel.py will tunnel a TCP connection through an HTTP server at PROXY_URL to SERVICE_URL.
+
+In Determined, this is used to tunnel ssh connections through the master, where the hostname in the
+SERVICE_URL should be the shell ID of the shell in question.
+"""
+
+
+class Copier(threading.Thread):
+    """
+    A thread to copy from one file descriptor to another.  Only copies in one direction; use two
+    threads to deal with bidirectional file descriptors.  The choice to use a pair of threads as
+    opposed to select.select or select.poll ensures that this code will run nicely on Windows.
+    """
+
+    def __init__(self, src: int, dst: int):
+        super().__init__()
+        self.src = src
+        self.dst = dst
+        self.keep_going = True
+
+    def run(self) -> None:
+        try:
+            while self.keep_going:
+                try:
+                    buf = os.read(self.src, 4096)
+                except OSError:
+                    break
+                if len(buf) == 0:
+                    break
+                try:
+                    os.write(self.dst, buf)
+                except OSError:
+                    break
+        finally:
+            # We're ok with double-closing some sockets.
+            try:
+                os.close(self.src)
+            except OSError:
+                pass
+
+            try:
+                os.close(self.dst)
+            except OSError:
+                pass
+
+
+def parse_url(url: str) -> Tuple[str, int]:
+    """Parse a url, or a host:port-style address."""
+    parsed_url = urllib.parse.urlparse(url)
+    if parsed_url.scheme:
+        assert parsed_url.hostname is not None, "No hostname in url: {}".format(url)
+        return parsed_url.hostname, int(parsed_url.port or socket.getservbyname(parsed_url.scheme))
+
+    parsed_url = urllib.parse.urlparse("http://" + url)
+    assert parsed_url.scheme, "Unable to parse url: {}".format(url)
+    assert parsed_url.hostname is not None, "No hostname in url: {}".format(url)
+    return parsed_url.hostname, int(parsed_url.port or socket.getservbyname(parsed_url.scheme))
+
+
+def http_connect_tunnel(proxy: str, service: str) -> None:
+    proxy_host, proxy_port = parse_url(proxy)
+    service_host, service_port = parse_url(service)
+
+    client = http.client.HTTPConnection(proxy_host, proxy_port)
+    client.set_tunnel(service_host, service_port)
+
+    try:
+        client.connect()
+    except socket.gaierror:
```

Comment:
non-blocking: Wait, can this even happen? Won't the CLI already have tried to look up the same host and failed by now?

Reply:
I think I hit this because of a bug in `tunnel.py` but the error was so cryptic, I decided it would be much easier to troubleshoot down the road if users who hit issues have a clear error message to give to us (or solve on their own).